### PR TITLE
Feature/feature crud weight

### PR DESCRIPTION
**Pull request by oricat8**

### CODE REVIEW

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

**Review by marclll — APPROVED** [review]

File: src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     * @param owner Username of the owner of the pet
+     * @param petName Name of the pet
+     * @param weightEntity The parameters of weight
+     */
+    @PostMapping("/{owner}/{petName}/{petDate}")
+    public void createWeight(@PathVariable String owner, @PathVariable String petName, @PathVariable String petDate,
+                             @RequestBody WeightEntity weightEntity) {
+        weightService.createWeight(owner, petName, petDate, weightEntity);
+    }
+    /**
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java:38]: All methods should be separated by one empty line

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+public interface WeightDao {
+    /**
+     * Creates a weight of the pet on the database.
+     * @param owner Username of the owner of the pet
+     * @param petName Name of the pet
+     * @param weightEntity The parameters of weight
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java:15]: Missing date param tag

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Repository
+public class WeightDaoImpl implements WeightDao {
+    private final String DELFAIL_KEY;
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Constants should be static final and be initialized inline

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        INVALID_WEIGHT_EXC = "invalid-pet";
+        internalListString1 = "date";
+        internalListString2 = "body";
+    }
+
+    @Override
+    public void createWeight(String owner, String petName, String date, WeightEntity weightEntity) {
+        DocumentReference weightRef = getWeightsRef(owner, petName).document(date);
+        weightRef.set(weightEntity);
+        System.out.println("Weight created");
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Prints must be deleted before merge

File: src/main/java/org/pesmypetcare/webservice/entity/WeightEntity.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package org.pesmypetcare.webservice.entity;
+
+import lombok.Data;
+
+@Data
+public class WeightEntity {
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/entity/WeightEntity.java:6]: Is it really necessary to have a class with just one attribute? Wouldn't it be better to use a variable instead?

File: src/main/java/org/pesmypetcare/webservice/service/WeightService.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+public interface WeightService {
+
+    /**
+     * Creates a weight of the pet on the database.
+     * @param owner Username of the owner of the pet
+     * @param petName Name of the pet
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/service/WeightService.java:14]: Missing date param tag

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestController.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@ExtendWith(MockitoExtension.class)
+public class PetWeightRestController {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestController.java]: The test class name should end with Test

**Review by santidrj — CHANGES_REQUESTED** [review]
There are some issues that need to be solved

**Review by marclll — CHANGES_REQUESTED** [review]
Fix Santiago's requests

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

File: src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java (hunk 1 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java:38]: Solved!

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java (hunk 2 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java:15]: Solved!

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java (hunk 3 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Solved!

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java (hunk 4 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Solved!

File: src/main/java/org/pesmypetcare/webservice/entity/WeightEntity.java (hunk 5 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/entity/WeightEntity.java:6]: To solve, we are going to fix it in the next iteration.

File: src/main/java/org/pesmypetcare/webservice/service/WeightService.java (hunk 6 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/service/WeightService.java:14]: Solved!

File: src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestController.java (hunk 7 above)
Comment (oricat8) [reply src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestController.java]: Solved!

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    public void updateWeight(String owner, String petName, String petDate, Object value) {
+        CollectionReference weightsRef = getWeightsRef(owner, petName);
+        DocumentReference weightRef = weightsRef.document(petDate);
+        weightRef.update("weightValue", value);
+    }
+
+    public CollectionReference getWeightsRef(String owner, String petName) {
+        return db.collection("users").document(owner).collection("pets").document(petName)
+            .collection("weights");
+    }
```
Comment (marclll) [inline src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: All functions have to be commented

**Review by marclll — CHANGES_REQUESTED** [review]

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java (hunk 8 above)
Comment (oricat8) [reply src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Solved!

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

File: src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Repository
+public class WeightDaoImpl implements WeightDao {
+    private static final String DELFAIL_KEY = "deletion-failed";
+    private static final String WEIGHT_DOES_NOT_EXIST_EXC = "The weight does not exist";
+    private static final String INVALID_WEIGHT_EXC = "invalid-pet";
+    private static final String internalListString1 = "date";
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/WeightDaoImpl.java]: Constants should be in upper case and each word separated by a low bar

**Review by santidrj — CHANGES_REQUESTED** [review]
There are some small changes to be done

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

**Review by santidrj — APPROVED** [review]
Everything seems ok now 👍

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

**Review by marclll — APPROVED** [review]
Well done!

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/service/WeightServiceImpl.java  1
- src/main/java/org/pesmypetcare/webservice/controller/petmanager/WeightRestController.java  1
- src/test/java/org/pesmypetcare/webservice/controller/petmanager/PetWeightRestControllerTest.java  1
         

Clones added
============
- src/main/java/org/pesmypetcare/webservice/service/WeightService.java  1
- src/main/java/org/pesmypetcare/webservice/dao/WeightDao.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5361118&bid=17566338)

**Pull request merged**